index.mjs

```diff
@@ -0,0 +1,165 @@
+import {URL, pathToFileURL} from 'node:url'
+
+import {unified} from 'unified'
+import {engine} from 'unified-engine'
+import {VFile} from 'vfile'
+import {
+  createConnection,
+  Diagnostic,
+  DiagnosticSeverity,
+  Position,
+  ProposedFeatures,
+  Range,
+  TextDocuments,
+  TextDocumentSyncKind
+} from 'vscode-languageserver/node.js'
+import {TextDocument} from 'vscode-languageserver-textdocument'
+
+/**
+ * Convert a unist point to a language server protocol position.
+ *
+ * @param {import('unist').Point} point
+ * @returns {Position}
+ */
+function unistPointToLSPPosition(point) {
+  return Position.create(point.line - 1, point.column - 1)
+}
+
+/**
+ * Convert a unist position to a language server protocol range.
+ *
+ * If no position is given, a range is returned  which represents the beginning
+ * of the document.
+ *
+ * @param {import('unist').Position?} position
+ * @returns {Range}
+ */
+function unistLocationToLSPRange(position) {
+  if (!position) {
+    return Range.create(0, 0, 0, 0)
+  }
+
+  const start = unistPointToLSPPosition(position.start)
+
+  return Range.create(
+    start,
+    // Fall back to start if the end position contains null values
+    position.end.line && position.end.column
+      ? unistPointToLSPPosition(position.end)
+      : start
+  )
+}
+
+/**
+ * Convert a vfile message to a language server protocol diagnostic.
+ *
+ * @param {import('vfile-message').VFileMessage} message
+ * @param {string|undefined} source
+ * @returns {Diagnostic}
+ */
+function vfileMessageToDiagnostic(message, source) {
+  const diagnostic = Diagnostic.create(
+    unistLocationToLSPRange(message.position),
+    message.reason,
+    message.fatal ? DiagnosticSeverity.Error : DiagnosticSeverity.Warning,
+    message.ruleId
+      ? message.source
+        ? message.source + ':' + message.ruleId
+        : message.ruleId
+      : message.source || undefined,
+    source
```

Comment:
Currently a hardcoded `source` is provided. The `code` provided is rendered the same as `vfile-reporter` renders it.

`vim-lsp` renders this as:

```
{filename}|{line} col {column} {level}| {source}:{level}:{code}:{message}
```

VS Code renders this as can be seen in this screenshow of an ESLint violation:

![image](https://user-images.githubusercontent.com/779047/145411109-673de6cc-777b-4b59-9b6c-4eba19884c03.png)

So given the warning in the example screenshot, VS Code would render it as:

```
Line must be at most 80 characters remark(remark-lint:maximum-line-length) [5, 21]
```

Alternatively I could provide the message source as the diagnostic `source`, perhaps with a fallback.

In this case `vim-lsp` would render:

```
README.md|7 col 109 warning| remark-lint:Warning:maximum-line-length:Line must be at most 80 characters
```

and VS Code would render:

```
Line must be at most 80 characters remark-lint(maximum-line-length) [5, 21]
```

Reply:
Where is the `remark` in `remark(remark-lint:maximum-line-length)` coming from? This is a unified project, right?

Not a VSC user, so maybe it’s normal, but `remark(remark-lint:maximum-line-length)` looks a bit weird to me? I’d think `remark-lint:maximum-line-length` is prettier?

And for the `vim-lsp` format, `remark-lint:Warning:maximum-line-length:Line`, seems weird that `remark-lint` and `maximum-line-length` are separated, and the casing (and placing) of `Warning`?

Reply:
Currently the `remark` is coming from `prefix`. How editors format is is more of an indication of what the properties mean. This is out of control of the language server.

I think typically the `source` is a hardcoded value, but for unified it makes sense to provide it dynamically. I’ll just use `message.source` for that with a fallback on `prefix` (or a dedicated property because of discussions below).